### ISSUE TITLE
refactor: GhostDrafter trait + speculative protocol corrections (ADR-0006)

### ISSUE THREAD

**Issue by SunayHegde2006** [SunayHegde2006/Air.rs#12]
## What to build

Introduce `GhostDrafter` trait seam for speculative decoding and apply 5 protocol corrections. See ADR-0006 (`docs/adr/0006-ghost-drafter-trait-speculative-corrections.md`) and `air_rs_speculative_decoding_protocol.md`.

### Protocol corrections (must ship with this issue)

**G1 — EOS inner break** (`src/speculative.rs`)
```rust
// In rejection_sample: break inside acceptance loop at EOS
accepted.push(token);
if token == eos_token { break; }
// Bonus token only when last accepted != EOS
```

**G2 — RNG type** (`src/speculative.rs`)
- `rng: rand::rngs::ThreadRng` → `rng: rand::rngs::SmallRng`
- Seeded: `SmallRng::from_rng(rand::thread_rng()).unwrap()`

**G3 — Draft sampler config** (`src/speculative.rs`)
- Add `sampler: SamplerConfig` to `SpeculativeConfig`
- Draft pass uses `self.config.sampler` — same as target

**G4 — `truncate_to` on `SessionKvCache`** (`src/kv_cache.rs`) *(amends ADR-0004)*
```rust
pub trait SessionKvCache: Send {
    // ... existing 4 methods ...
    fn truncate_to(&mut self, seq_len: usize);  // O(1) pointer rollback
}
// impl for KvCacheManager: self.seq_len_used[..] = seq_len
// impl for MockSessionKvCache: no-op
```

### GhostDrafter trait (new file: `src/ghost_drafter.rs`)

```rust
pub trait GhostDrafter: Send + Sync {
    fn draft_pass(&mut self, context: &[u32], k: usize, sampler: &SamplerConfig)
        -> Result<DraftResult>;
    fn on_accept(&mut self, n_accept: usize, context_len: usize);
    fn reset(&mut self);
}

// DraftResult: tokens: Vec<u32>, logits: Vec<Vec<f32>>

pub struct StreamingLayerSkipDrafter { /* draft_kv, manifest, streamer */ }
impl GhostDrafter for StreamingLayerSkipDrafter { ... }

#[cfg(test)]
pub struct MockDrafter { tokens: Vec<u32>, logits: Vec<Vec<f32>> }
#[cfg(test)]
impl GhostDrafter for MockDrafter { ... }
```

### `Speculative` struct changes (`src/speculative.rs`)

- `drafter: Box<dyn GhostDrafter>` replaces inline draft pass logic
- `target_kv: Box<dyn SessionKvCache>` (uses ADR-0004 seam)
- `rng: SmallRng`
- `Speculative::new(config, drafter: Box<dyn GhostDrafter>, gguf_path)` — drafter injected

### `stream_layers` primitive (`src/generator.rs`)

```rust
pub fn stream_layers(
    &mut self,
    layer_offsets: &[u64],
    input: &[u32],
    kv: &mut dyn SessionKvCache,
    device: &Device,
) -> Result<Tensor>
```
`generate_step` refactored to call this internally. `StreamingLayerSkipDrafter` uses it.

### `.gitignore`

- Add `*.draft.json` (sidecar manifests — never tracked)
- Add `!air_rs_speculative_decoding_protocol.md` (tracked)

## Acceptance criteria

- [ ] `SessionKvCache` has 5 methods including `truncate_to`; `KvCacheManager` implements it; `MockSessionKvCache` no-ops it
- [ ] `GhostDrafter` trait in `src/ghost_drafter.rs` with 3 methods
- [ ] `StreamingLayerSkipDrafter` implements `GhostDrafter`
- [ ] `#[cfg(test)] MockDrafter` in `ghost_drafter.rs` returns canned tokens
- [ ] `Speculative` holds `Box<dyn GhostDrafter>` — no inline draft logic
- [ ] `Speculative.rng` is `SmallRng` (Send-safe)
- [ ] `SpeculativeConfig.sampler: SamplerConfig` — draft and target use same params
- [ ] EOS inner break: tokens after EOS never emitted
- [ ] `stream_layers` exposed on `InferenceGenerator`
- [ ] Unit test: `MockDrafter` + `MockSessionKvCache` + `Device::Cpu` → full speculative round, no GPU
- [ ] Unit test: `rejection_sample` with synthetic logits matches distribution (Monte Carlo, 10k samples)
- [ ] Unit test: EOS in position 1 of 4 draft tokens → only 2 tokens emitted (t0=EOS + 0 after)
- [ ] `cargo check` zero errors, `cargo test` all passing

## Blocked by

#7 (TransformerBlock — `stream_layers` refactors alongside)
#10 (SessionKvCache — `truncate_to` amendment)

**Comment by SunayHegde2006** [comment]
Implemented in commit `2b94d06` (PR #13).

**Delivered:**
- `src/ghost_drafter.rs` — `GhostDrafter` trait: `draft_pass`, `on_accept`, `reset`
- `DraftResult`: token candidates + logprob arrays + accept mask
- `SamplerConfig`: shared draft/target temperature+top_p (Gap 3 fix — enforces Leviathan et al. 2023 Theorem 1)
- `SpeculativeConfig`: `gamma` + `SamplerConfig` bundle
- `MockDrafter`: deterministic test double
- Protocol: draft uses same `SamplerConfig` as target for valid rejection sampling
